app/gui/src/presenter/graph.rs

```diff
@@ -185,29 +180,27 @@ impl Model {
     ///
     /// The behavior of this function can be summarized in the following table:
     /// ```ignore
-    /// | Context Enabled | Active      | Action       |
-    /// |-----------------|-------------|--------------|
-    /// | Yes             | Yes         | Add Disable  |
-    /// | Yes             | No          | Clear        |
-    /// | No              | Yes         | Add Enable   |
-    /// | No              | No          | Clear        |
+    /// | Global Context Permission | Active      | Action       |
+    /// |---------------------------|-------------|--------------|
+    /// | Enabled                   | Yes         | Add Disable  |
+    /// | Enabled                   | No          | Clear        |
+    /// | Disabled                  | Yes         | Add Enable   |
+    /// | Disabled                  | No          | Clear        |
     /// ```
-    /// TODO(#5929): Connect this function with buttons on nodes.
-    #[allow(dead_code)]
     fn node_action_context_switch(&self, id: ViewNodeId, active: bool) {
         let context = Context::Output;
-        let current_state = self.output_context_enabled();
-        let environment = self.execution_environment().into();
+        let environment = self.execution_environment.get();
+        let current_state = environment.output_context_enabled();
         let switch = if current_state { ContextSwitch::Disable } else { ContextSwitch::Enable };
         let expr = if active {
-            Some(ContextSwitchExpression { switch, context, environment })
+            let environment_name = environment.to_string().into();
+            Some(ContextSwitchExpression { switch, context, environment: environment_name })
         } else {
             None
         };
         self.log_action(
             || {
-                let ast_id =
-                    self.state.update_from_view().set_node_context_switch(id, expr.clone())?;
+                let ast_id = self.state.ast_node_id_of_view(id)?;
```

Comment:
This is the part that I changed to get the expression updates to reach the view. Based on my understanding, what used to happen is this:
1. In this function (which triggers when the button is clicked), both the presenter state and the controller get updated.
2. When we get an update back from the controller, we trigger this FRP link: `expression_update <= update_data.map(|update| update.set_node_expressions());`.
3. In `set_node_expressions`, the nodes are checked for a mismatch between the controller update and the presenter state. However, since the state was already updated, no mismatch was found and therefore no update is triggered.

By removing the explicit view update here, we create that mismatch and the update fires correctly. I'm not sure if this was the correct approach. In any case, I think the skip and freeze changes have the same problem. Should I do the same thing there?

Reply:
I'm not sure. How to reproduce this issue in a previous version of the code? I don't understand why we need to update the view if it was already updated.